keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,428 @@
+   - `protocolSignature`: Specifies the protocol targeted by this Bucket instance.  One of:
+     - `azureBlob`: data required to target a provisioned azure container and/or storage account.
+     - `s3`: data required to target a provisioned S3 bucket and/or user.
+     - `gcs`: data required to target a provisioned GCS bucket and/or service account.
+1. `parameters`: a copy of the BucketClass parameters.
+1. `message`: a human readable description detailing the reason for the current `phase``.
+1. `phase`: is the current state of the Bucket:
+   - _Ready_: the controller finished processing and the controller is ready to receive access requests.
+   - _Released_: the originating Bucket has been deleted, signalling that the Bucket is ready for garbage collection.  This will occur on Greenfield Buckets once all requests referencing the Bucket are deleted.
+   - _Failed_: error and all retries have been exhausted.
+   - _Retrying_: set when a driver or Kubernetes error is encountered during provisioning operations indicating a retry loop.
+
+#### BucketClass
+
+A cluster-scoped API whose purpose it to provide admins control over the handling of bucket provisioning by specifying provisioner-defined key-values in the `parameters` map.  The `BucketClass` defines a release policy, and specifies driver specific parameters and the provisioner name.
+
+```yaml
+apiVersion: cosi.io/v1alpha1
+kind: BucketClass
+metadata:
+  name: 
+provisioner: [1]
+isDefaultBucketClass: [2]
+supportedProtocols: {"azureblob", "gcs", "s3", ... } [3]
+anonymousAccessMode: {"ro", "wo", "rw"} [4]
+additionalPermittedNamespaces: [5]
+- name:
+  uid: 
+releasePolicy: {"Delete", "Retain"} [6]
+parameters: [7]
+```
+
+1. `provisioner`: The name of the driver. If supplied the driver container and sidecar container are expected to be deployed. If omitted the `secretRef` is required for static provisioning.
+1. `isDefaultBucketClass`: boolean. When true, signals that the controller should attempt to match `Bucket`’s without a defined `BucketClass` to this class, accounting for the `Bucket`’s requested protocol.  Multiple default classes for the same protocol will produce undefined behaviour.
+1. `supportedProtocols`: protocols the associated object store supports.  Applied when matching Bucket to BucketClasses. Admins may specify all protocols a provisioner supports.  `Bucket.spec.protocol` will be checked against the array prior to provisioning.
+1. `anonymousAccessMode`: (Optional) ACL specifying *uncredentialed* access to the Bucket.  This is applicable for cases where the storage instance or objects are intended to be publicly readable and/or writable.
+1. `additionalPermittedNamespaces`: A list of namespace *in addition to the originating namespace* that will be allowed access to this Bucket.  A nil and empty string value are equivalent and must not append any namespace to `Bucket.spec.permittedNamespaces` list.
```

Comment:
@wlan0 Is this field needed to handle the _greenfield_ -> _brownfield_ shared use-case?

Reply:
_Maybe_ we should delete this field until we can provide solid justification. Easier to add a field than to remove (I think)...